### PR TITLE
Remove the experimental MiniLLM trainer

### LINKED ISSUE #6626: MiniLLM: `_compute_advantage` discounts by absolute position, scaling every advantage by gamma^t
### Reproduction

`MiniLLMTrainer._compute_advantage` builds `gamma_pow[i] = gamma**i` from the **absolute** index `i`, then does a reverse cumsum. That computes `sum_{i>=t} gamma^i * R_i = gamma^t * A_t` instead of the documented `sum_{i>=t} gamma^(i-t) * R_i = A_t`, so every position is scaled by an extra `gamma^t`.

**No exception is raised** — with `length_normalization=False` the advantages are silently wrong (later tokens are exponentially suppressed toward zero). With `length_normalization=True` the `gamma^t` cancels between numerator and denominator, but `gamma**i` underflows to exactly `0.0` in float32, giving `0/0 = nan` on long completions.

The repro calls the method with a stub `self` because it reads only `self.gamma` and `self.length_normalization`.

```python
import types

import torch

from trl.experimental.minillm.minillm_trainer import MiniLLMTrainer

advantage = MiniLLMTrainer._compute_advantage  # real method, called with a stub `self`

# Constant reward of 1.0 at every position: rewards = teacher - student = 1 - 0 = 1
T, GAMMA = 512, 0.9
student, teacher, mask = torch.zeros(1, T), torch.ones(1, T), torch.ones(1, T)

print("BUG 1 (length_normalization=False): every position is scaled by an extra gamma^t.")
print(f"With a constant reward of 1.0, the advantage should be A_t = (1-{GAMMA}^(T-t))/(1-{GAMMA}) = 10.0")
print("at every position. Instead it decays, and the error is exactly a factor of gamma^t:\n")
out = advantage(types.SimpleNamespace(gamma=GAMMA, length_normalization=False), student, teacher, mask)[0]
print(f"{'t':>5} {'expected':>12} {'actual':>14} {'actual/expected':>17} {'gamma^t':>14}")
for t in [0, 50, 100, 200, 400]:
    expected = (1 - GAMMA ** (T - t)) / (1 - GAMMA)
    print(f"{t:>5} {expected:>12.4f} {out[t].item():>14.6e} {out[t].item() / expected:>17.6e} {GAMMA**t:>14.6e}")
print("\n^ the last two columns match: gamma_pow is indexed by the absolute i, not by (i - t).\n")

print("BUG 2 (length_normalization=True): the gamma^t factor cancels between numerator and")
print("denominator, so every position should be exactly 1.0 -- until gamma^i underflows to 0.0")
print("in float32. Past that point both sums are 0.0 and the division yields 0/0 = nan:\n")
for gamma, T2 in [(0.5, 512), (0.9, 1024), (0.95, 2048), (0.99, 16384)]:
    out = advantage(types.SimpleNamespace(gamma=gamma, length_normalization=True), torch.zeros(1, T2),
                    torch.ones(1, T2), torch.ones(1, T2))[0]
    nan_at = (~torch.isfinite(out)).nonzero()
    zero_at = (torch.pow(gamma, torch.arange(T2)) == 0.0).nonzero()
    print(f"  gamma={gamma:<5} T={T2:<5} A_0={out[0].item():<5} A_last={out[-1].item():<5} "
          f"first nan at t={str(int(nan_at[0, 0])) if len(nan_at) else 'never':<7} "
          f"first gamma^i == 0.0 at i={str(int(zero_at[0, 0])) if len(zero_at) else 'never'}")
print("\n^ the nan index always equals the underflow index. Smaller gamma underflows sooner.")
```

outputs:

```
BUG 1 (length_normalization=False): every position is scaled by an extra gamma^t.
With a constant reward of 1.0, the advantage should be A_t = (1-0.9^(T-t))/(1-0.9) = 10.0
at every position. Instead it decays, and the error is exactly a factor of gamma^t:

    t     expected         actual   actual/expected        gamma^t
    0      10.0000   9.999997e+00      9.999997e-01   1.000000e+00
   50      10.0000   5.153767e-02      5.153767e-03   5.153775e-03
  100      10.0000   2.656132e-04      2.656132e-05   2.656140e-05
  200      10.0000   7.055040e-09      7.055040e-10   7.055079e-10
  400       9.9999   4.977323e-18      4.977360e-19   4.977414e-19

^ the last two columns match: gamma_pow is indexed by the absolute i, not by (i - t).

BUG 2 (length_normalization=True): the gamma^t factor cancels between numerator and
denominator, so every position should be exactly 1.0 -- until gamma^i underflows to 0.0
in float32. Past that point both sums are 0.0 and the division yields 0/0 = nan:

  gamma=0.5   T=512   A_0=1.0   A_last=nan   first nan at t=150     first gamma^i == 0.0 at i=150
  gamma=0.9   T=1024  A_0=1.0   A_last=nan   first nan at t=987     first gamma^i == 0.0 at i=987
  gamma=0.95  T=2048  A_0=1.0   A_last=nan   first nan at t=2028    first gamma^i == 0.0 at i=2028
  gamma=0.99  T=16384 A_0=1.0   A_last=nan   first nan at t=10346   first gamma^i == 0.0 at i=10346

^ the nan index always equals the underflow index. Smaller gamma underflows sooner.
```

The referenced implementation ([LMOps `losses.py#L37-L49`](https://github.com/microsoft/LMOps/blob/e210d2c026b9958617887762400778ace81172e6/minillm/minillm/losses.py#L37-L49)) uses a backward recurrence, `last_rw = rewards[:, t] + gamma * last_rw`, which produces the relative exponent by construction and never materializes `gamma**i` — so it has neither problem. Porting that recurrence would fix both at once.

Happy to submit a PR porting the backward recurrence.

### System Info

```
- Platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
- Python version: 3.12.4
- TRL version: 1.10.0.dev0+aa59202
- PyTorch version: 2.13.0
- accelerator(s): NVIDIA GeForce RTX 4090
- Transformers version: 5.14.1
- Accelerate version: 1.14.0
- Accelerate config: not found
- Datasets version: 5.0.1
- HF Hub version: 1.26.0
- bitsandbytes version: 0.50.0
- DeepSpeed version: 0.19.3
- Liger-Kernel version: 0.8.0
- PEFT version: 0.20.0
- vLLM version: not installed
```

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### LINKED ISSUE #7295: Proposing to remove MiniLLMTrainer
Proposing to remove `trl.experimental.minillm` (`MiniLLMTrainer`, `MiniLLMConfig`), added in #4504 (2025-11-12), directly into `trl.experimental`. Reporting the inputs discussed in #7182.

**Usage.** 
- 13 trainings in the last 30 days as per our telemetry. 
- 9 models carry both the `minillm` and `trl` tags, 8 of them from two accounts in one sweep in May: https://huggingface.co/models?other=minillm,trl&sort=created. 
- Nobody has spoken up for it in #7182.

**External issues and pull requests.** 
- 3 issues and 6 pull requests from outside the maintainer team in the last twelve months. 
- Two issues are real bugs in the advantage computation: #6626 (discount by absolute position, fix in #6635, open since August) and #7024 (padding leaking into the advantage, closed). 
- #4743 from the paper author has been stale since March. The rest are doc link fixes and CLI JSON parsing.

**Subsumed.** 
- Partially. The `rkl_advantage=False, single_step_decomposition=True` variant is the token-level reverse KL that `DistillationTrainer(beta=1.0)` already trains on-policy. What goes away is the sequence-level reverse KL optimized as an RL reward with `gamma` discounting, which is where both open bugs live.

**Cost.** 
- 567 lines of trainer and config, 716 with test and doc page. 
- It duplicates the GRPO generation path

**Downstream consumers.** 
- None. GitHub code search finds three zero-star personal repos importing it, everything else is vendored copies of the trl tree.

**Owner.** 
- None has come forward.

**Age.** 
- 10 months in `trl.experimental`, shipped in every release from v0.26.0 to v1.13.0 with only upkeep commits since it landed. Long enough to have been found.

This issues aims to collect feedback from the community if any. If none:
- remove the module, test and doc page, 
- keep the paper index entry pointing to the git history, and
- close #6626, #6635 and #4743 as part of it.

### PR DESCRIPTION
**Pull request by yashb98**
# What does this PR do?

I've prepared the removal proposed in #7295. This removes the experimental MiniLLM package, its tests and documentation page, and keeps the paper index entry with the last supported release noted. The maintainer decision and timing are still pending in that issue.

I kept the shared GRPO and GSPO-token logic for per-token advantages unchanged, removing only the MiniLLM example from their comments. The telemetry entry and documentation navigation references are also removed.

Fixes #7295. Closes #6626. If the removal goes ahead, the related PRs #6635 and #4743 can also be closed by the maintainers. PR #6573 touches the MiniLLM trainer too, so the changes will need coordination if it lands first.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Discussion: #7295; removal approval is still pending.
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? No new tests for this removal; preserved behavior was checked below.

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Testing

- `make precommit` passes, including copyright checks, Ruff and doc-builder-style on all files, without changing the source.
- `tests/test_grpo_trainer.py::TestGRPOTrainer::test_train_loss_types`: 7 passed and 7 skipped on both the baseline and this commit. The skips require liger-kernel, which is not installed locally.
- Experimental test collection changes from 539 to 534, exactly the five removed MiniLLM tests. GRPO still collects 202 tests on both revisions.
- TRL, experimental, GRPO and GSPO-token imports pass. MiniLLM is absent as expected; only the retained paper entry mentions it.
- Local MDX generation and internal-link resolution pass with the pinned doc-builder source.

Full main/experimental suites, Liger cases and HTML rendering have not run. Remote CI is pending.

## Who can review?

@qgallouedec

---

> [!NOTE]
> **Medium Risk**
> This is a breaking API removal for anyone importing `trl.experimental.minillm`; runtime GRPO behavior is unchanged aside from comment edits.
> 
> **Overview**
> Removes the experimental **MiniLLM** distillation stack (`MiniLLMTrainer`, `MiniLLMConfig`, package under `trl/experimental/minillm/`) along with its tests and dedicated documentation page.
> 
> Documentation is updated so the experimental trainer list and navigation no longer link to MiniLLM; the paper index entry for *Knowledge Distillation of Large Language Models* now states that `MiniLLMTrainer` was available through TRL v1.13 and lives only in git history. **`MiniLLMTrainer`** is dropped from the telemetry trainer allowlist in `base_trainer.py`.
> 
> GRPO and GSPO-token **per-token advantage** handling is unchanged; comments no longer cite MiniLLM as the motivating example.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 19da0dd72d9fe8f30ba7ad7dd966bfe0d34cc7f2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
Closing: we don't review fully AI-generated PRs from first-time contributors, see the [AI usage policy](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#ai-usage-policy).

**Comment by yashb98** [comment]
> Closing: we don't review fully AI-generated PRs from first-time contributors, see the [AI usage policy](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#ai-usage-policy).

Understood. Would you consider a fresh PR for this removal if I implement the changes and run the required checks myself? I'd like to follow your policy and will wait for your guidance before resubmitting.
